test/e2e/azure_securitygroups.go

```diff
@@ -0,0 +1,188 @@
+	Expect(mgmtClient).NotTo(BeNil())
+
+	// get subscription id
+	settings, err := auth.GetSettingsFromEnvironment()
+	Expect(err).NotTo(HaveOccurred())
+	subscriptionID := settings.GetSubscriptionID()
+	auth, err := azureutil.GetAuthorizer(settings)
+	Expect(err).NotTo(HaveOccurred())
+
+	By("creating a subnets client")
+	subnetsClient := network.NewSubnetsClient(subscriptionID)
+	subnetsClient.Authorizer = auth
+
+	securityGroupsClient := network.NewSecurityGroupsClient(subscriptionID)
+	securityGroupsClient.Authorizer = auth
+
+	securityRulesClient := network.NewSecurityRulesClient(subscriptionID)
+	securityRulesClient.Authorizer = auth
+
+	azureCluster := &infrav1.AzureCluster{}
+	err = mgmtClient.Get(ctx, client.ObjectKey{
+		Namespace: input.Cluster.Spec.InfrastructureRef.Namespace,
+		Name:      input.Cluster.Spec.InfrastructureRef.Name,
+	}, azureCluster)
+	Expect(err).NotTo(HaveOccurred())
+
+	var expectedSubnets infrav1.Subnets
+	checkSubnets := func(g Gomega) {
+		for _, expectedSubnet := range expectedSubnets {
+			securityGroup, err := securityGroupsClient.Get(ctx, azureCluster.Spec.ResourceGroup, expectedSubnet.SecurityGroup.Name, "")
+			g.Expect(err).NotTo(HaveOccurred())
+
+			securityRules, err := securityRulesClient.List(ctx, azureCluster.Spec.ResourceGroup, *securityGroup.Name)
+			g.Expect(err).NotTo(HaveOccurred())
+
+			var expectedSecurityRuleNames []string
+			for _, expectedSecurityRule := range expectedSubnet.SecurityGroup.SecurityRules {
+				expectedSecurityRuleNames = append(expectedSecurityRuleNames, expectedSecurityRule.Name)
+			}
+
+			for _, securityRule := range securityRules.Values() {
+				g.Expect(expectedSecurityRuleNames).To(ContainElement(*securityRule.Name))
+			}
+		}
+	}
+
+	Byf("Creating subnets for the %s cluster", input.ClusterName)
+	testSubnet := infrav1.SubnetSpec{
+		SubnetClassSpec: infrav1.SubnetClassSpec{
+			Name: "test-subnet",
+			Role: infrav1.SubnetNode,
+		},
+		SecurityGroup: infrav1.SecurityGroup{
+			Name: "test-security-group",
+			SecurityGroupClass: infrav1.SecurityGroupClass{
+				SecurityRules: []infrav1.SecurityRule{
+					testSecurityRule,
+				},
+			},
+		},
+	}
+	originalSubnets := azureCluster.Spec.NetworkSpec.Subnets
+	expectedSubnets = originalSubnets
+	expectedSubnets = append(expectedSubnets, testSubnet)
+	Eventually(func(g Gomega) {
+		g.Expect(mgmtClient.Get(ctx, client.ObjectKeyFromObject(azureCluster), azureCluster)).To(Succeed())
+		azureCluster.Spec.NetworkSpec.Subnets = expectedSubnets
+		g.Expect(mgmtClient.Update(ctx, azureCluster)).To(Succeed())
+	}, inputGetter().WaitForUpdate...).Should(Succeed())
+	Eventually(checkSubnets, input.WaitForUpdate...).Should(Succeed())
+
+	By("Creating new security rule for the subnet")
+	Expect(len(expectedSubnets)).To(Not(Equal(0)))
+	testSubnet.SecurityGroup.SecurityRules = infrav1.SecurityRules{testSecurityRule, testSecurityRule2}
+	expectedSubnets = originalSubnets
+	expectedSubnets = append(expectedSubnets, testSubnet)
```

Comment:
nit: could this just be `expectedSubnets = append(originalSubnets, testSubnet)` so we can get rid of the line above?

Reply:
The linter complains saying that `append` must be used in this way: `foo = append(foo, bar)`